spec/03-plugins/10-key-auth/02-access_spec.lua

```diff
@@ -347,22 +362,49 @@ describe("Plugin: key-auth (access)", function()
 
     for type, _ in pairs(harness) do
       describe(type, function()
-        it("false sends key to upstream", function()
-          local res   = assert(client:send(harness[type][1]))
-          local body  = assert.res_status(200, res)
-          local json  = cjson.decode(body)
-          local field = type == "post_data.params" and json.post_data.params or json[type]
-          assert.equal("kong", field.apikey)
-        end)
-        it("true doesn't send key to upstream", function()
-          local res   = assert(client:send(harness[type][2]))
-          local body  = assert.res_status(200, res)
-          local json  = cjson.decode(body)
-          local field = type == "post_data.params" and json.post_data.params or json[type]
-          assert.is_nil(field.apikey)
-        end)
+        for _, content_type in pairs({
+          "application/x-www-form-urlencoded",
+          "application/json",
+          "multipart/form-data",
+          }) do
+
+          if type == "post_data.params" then
+            harness[type][1].headers["Content-Type"] = content_type
+            harness[type][2].headers["Content-Type"] = content_type
+          end
+
+          it("(" .. content_type .. ") false sends key to upstream", function()
+            local res   = assert(client:send(harness[type][1]))
+            local body  = assert.res_status(200, res)
+            local json  = cjson.decode(body)
+            local field = type == "post_data.params" and json.post_data.params or json[type]
+            assert.equal("kong", field.apikey)
+          end)
+          it("(" .. content_type .. ") true doesn't send key to upstream", function()
+            local res   = assert(client:send(harness[type][2]))
+            local body  = assert.res_status(200, res)
+            local json  = cjson.decode(body)
+            local field = type == "post_data.params" and json.post_data.params or json[type]
+            assert.is_nil(field.apikey)
+          end)
+        end
```

Comment:
I am of the opinion that we should add a test proving that the body's MIME type has not been tampered with. Considering the above comment wrt. mutating requests, I propose such a patch to introduce this test:
```diff
diff --git a/spec/03-plugins/10-key-auth/02-access_spec.lua b/spec/03-plugins/10-key-auth/02-access_spec.lua
index 3095e698..f076d5da 100644
--- a/spec/03-plugins/10-key-auth/02-access_spec.lua
+++ b/spec/03-plugins/10-key-auth/02-access_spec.lua
@@ -319,54 +319,55 @@ describe("Plugin: key-auth (access)", function()
 
   describe("config.hide_credentials", function()
 
-    local harness = {
-      uri_args = { -- query string
-        {
-          headers = { Host = "key-auth1.com" },
-          path    = "/request?apikey=kong",
-          method  = "GET",
+    for _, content_type in pairs({
+      "application/x-www-form-urlencoded",
+      "application/json",
+      "multipart/form-data",
+      }) do
+
+      local harness = {
+        uri_args = { -- query string
+          {
+            headers = { Host = "key-auth1.com" },
+            path    = "/request?apikey=kong",
+            method  = "GET",
+          },
+          {
+            headers = { Host = "key-auth2.com" },
+            path    = "/request?apikey=kong",
+            method  = "GET",
+          }
         },
-        {
-          headers = { Host = "key-auth2.com" },
-          path    = "/request?apikey=kong",
-          method  = "GET",
-        }
-      },
-      headers = {
-        {
-          headers = { Host = "key-auth1.com", ["apikey"] = "kong" },
-          path    = "/request",
-          method  = "GET",
-        },
-        {
-          headers = { Host = "key-auth2.com", ["apikey"] = "kong" },
-          path    = "/request",
-          method  = "GET",
-        },
-      },
-      ["post_data.params"] = {
-        {
-          headers = { ["Host"] = "key-auth5.com", },
-          body    = { apikey = "kong" },
-          method  = "POST",
-          path    = "/request",
-        },
-        {
-          headers = { ["Host"] = "key-auth6.com", },
-          body    = { apikey = "kong" },
-          method  = "POST",
-          path    = "/request",
+        headers = {
+          {
+            headers = { Host = "key-auth1.com", ["apikey"] = "kong" },
+            path    = "/request",
+            method  = "GET",
+          },
+          {
+            headers = { Host = "key-auth2.com", ["apikey"] = "kong" },
+            path    = "/request",
+            method  = "GET",
+          },
         },
+        ["post_data.params"] = {
+          {
+            headers = { ["Host"] = "key-auth5.com", },
+            body    = { apikey = "kong" },
+            method  = "POST",
+            path    = "/request",
+          },
+          {
+            headers = { ["Host"] = "key-auth6.com", },
+            body    = { apikey = "kong", foo = "bar" },
+            method  = "POST",
+            path    = "/request",
+          },
+        }
       }
-    }
 
-    for type, _ in pairs(harness) do
-      describe(type, function()
-        for _, content_type in pairs({
-          "application/x-www-form-urlencoded",
-          "application/json",
-          "multipart/form-data",
-          }) do
+      for type, _ in pairs(harness) do
+        describe(type, function()
 
           if type == "post_data.params" then
             harness[type][1].headers["Content-Type"] = content_type
@@ -387,7 +388,17 @@ describe("Plugin: key-auth (access)", function()
             local field = type == "post_data.params" and json.post_data.params or json[type]
             assert.is_nil(field.apikey)
           end)
-        end
+        end)
+      end
+
+      it("(" .. content_type .. ") true preserves body MIME type", function()
+        local req = harness["post_data.params"][2]
+        req.headers["Content-Type"] = content_type
+
+        local res   = assert(client:send(req))
+        local body  = assert.res_status(200, res)
+        local json  = cjson.decode(body)
+        assert.equal("bar", json.post_data.params.foo)
       end)
     end
```

I am currently hitting an exception for the multipart encoding with lua-multipart 0.5.4 which I have not investigated, but pinged @bungle (here is the test-case to reproduce, with the above body `{ apikey = "kong", foo = "bar" }`)